internals/src/error.rs

```diff
@@ -26,3 +26,27 @@ macro_rules! write_err {
         }
     }
 }
+
+/// Trait used internally to assert that error types implement `Send`, `Sync`, and `Error`.
+#[cfg(all(test, feature = "std"))]
+#[doc(hidden)]
+pub trait AssertTraits: Send + Sync + std::error::Error {}
```

Comment:
Still not super happy this is pub.

I would suggest the following changes:
- have a single macro and have a `($type:ty) => {};` rule for non-test, non-empty rule for test
- do Sanket's approach for methods, have one method `fn check_sendsync<T: Send + Sync>()` and `fn check_error<T: std::error::Error>()` and gate the second on std.

Currently there is a bug that non-std test runs will not be able to build.

Reply:
I'm not happy about it either but seems less work for compiler (and compile times!). It's hidden which is a clear signal people shouldn't use it (and internals is also such signal). If anyone is insane enough to depend on it they deserve the consequences.

Reply:
I used you suggestions @stevenroose except I was unable to put an attribute on the arms of the macro so I kept the dummy macro for not test builds.